bundles/org.smarthomej.binding.viessmann/src/main/java/org/smarthomej/binding/viessmann/internal/handler/DeviceHandler.java

```diff
@@ -226,169 +230,253 @@ public void handleUpdate(FeatureDataDTO featureDataDTO) {
+                                case "year":
+                                    subMsg.setSuffix("thisYear");
+                                    subMsg.setChannelType("type-energy");
+                                    createSubChannel(subMsg);
+                                    subMsg.setSuffix("lastYear");
+                                    createSubChannel(subMsg);
+                                    break;
+                                default:
+                                    break;
                             }
-                        }
 
-                        if ("temperature".equals(typeEntry)) {
-                            DecimalType state = DecimalType.valueOf(msg.getValue());
-                            updateState(msg.getChannelId(), state);
-                        } else if ("number".equals(typeEntry)) {
-                            DecimalType state = DecimalType.valueOf(msg.getValue());
-                            updateState(msg.getChannelId(), state);
-                        } else if ("boolean".equals(typeEntry)) {
-                            OnOffType state = bool ? OnOffType.ON : OnOffType.OFF;
-                            updateState(msg.getChannelId(), state);
-                        } else if ("string".equals(typeEntry) || "array".equals(typeEntry)) {
-                            StringType state = StringType.valueOf(msg.getValue());
-                            updateState(msg.getChannelId(), state);
-                        } else if ("Schedule".equals(typeEntry)) {
-                            StringType state = StringType.valueOf(msg.getValue());
-                            updateState(msg.getChannelId(), state);
-                            String channelId = msg.getChannelId().replace("#schedule", "#produced");
-                            updateState(channelId, parseSchedule(msg.getValue()));
+                            switch (typeEntry) {
+                                case "decimal":
+                                case "number":
+                                case "temperature":
+                                    if (unit != null) {
+                                        if (unit.isEmpty()) {
+                                            DecimalType state = DecimalType.valueOf(msg.getValue());
+                                            updateState(msg.getChannelId(), state);
+                                        } else {
+                                            updateState(msg.getChannelId(),
+                                                    new QuantityType<>(msg.getValue() + " " + unit));
+                                        }
+                                    }
+                                    break;
+                                case "percent":
+                                    updateState(msg.getChannelId(),
+                                            new QuantityType<>(Double.valueOf(msg.getValue()), Units.PERCENT));
```

Comment:
Also below: Why don't you use the unit map for that, too? Then you can just have one code for that instead of many cases.